src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -6605,6 +6570,98 @@ public override IOperation VisitRangeOperation(IRangeOperation operation, int? a
             return new RangeOperation(operation.IsLifted, operation.IsImplicit, semanticModel: null, operation.Syntax, operation.Type, visitedLeftOperand, visitedRightOperand, operation.Method);
         }
 
+        public override IOperation VisitSwitchExpression(ISwitchExpressionOperation operation, int? captureIdForResult)
+        {
+            // PROTOTYPE(recusrive-patterns): This is known not to be correct or working.
+
+            // expression switch { pat1 when g1 => e1, pat2 when g2 => e2 }
+            //
+            // becomes
+            //
+            // captureInput = expression
+            // START scope 1 (arm1 locals)
+            // GotoIfFalse (captureInput is pat1 && g1) label1;
+            // captureOutput = e1
+            // goto afterSwitch
+            // label1:
+            // END scope 1
+            // START scope 2
+            // GotoIfFalse (captureInput is pat2 && g2) label2;
+            // captureOutput = e2
+            // goto afterSwitch
+            // label2:
+            // END scope 2
+            // throw new switch failure
+            // afterSwitch:
+            // result = captureOutput
+
+            INamedTypeSymbol booleanType = _compilation.GetSpecialType(SpecialType.System_Boolean);
+            SpillEvalStack();
+            var capturedInput = VisitAndCapture(operation.Value);
+            int captureOutput = captureIdForResult ?? GetNextCaptureId(_currentRegion);
```

Comment:
>int captureOutput = captureIdForResult ?? GetNextCaptureId(_currentRegion); [](start = 12, length = 75)

This doesn't feel right. The previous statement is likely to change the ```_currentRegion```, but we want the captureOutput to be owned by the region that was active when we entered the function. When we are done we should leave all regions up that initial region, all other "scopes", including the one for the ```capturedInput``` should be "closed".

Reply:
Reordered as suggested, but it appears this misordering issue may exist in many other places (e.g. in `VisitDynamicBinaryConditionalOperator`, `VisitUserDefinedBinaryConditionalOperator`, etc)

---
In reply to: [244425546](https://github.com/dotnet/roslyn/pull/32018#discussion_r244425546) [](ancestors = 244425546)